### PR TITLE
refactor: rewrite imagetool to improve interactions and accessibility

### PR DESCRIPTION
**Pull request by jmswrnr**
### Description

This PR includes significant changes to the hotspot/crop `ImageToolInput` component and its related files. The primary focus is on refactoring the code to improve interactions, accessibility, and maintainability, and replacing the canvas-based implementation with an SVG-based one.

This new implementation includes rewriting of component logic which came with some improvements:
- Adds focus and hover states to hotspot and crop area
- Improved interaction logic:
  - The hotspot area doesn't exceed crop bounds in previews while dragging.
  - Prevents resizing hotspot area into negative/inverted size
  - Prevents resizing crop area too small which can get stuck
  - Switched from event-based delta movement for dragging to a delta relative to initial event origin. Similar to modern image editing tools, this means you must return the cursor to original position before dragging the other direction and can confidently drag the element to the edge/corner of the bounds.
- Adds keyboard navigation (in separate commit)

### What to review

This replaces the UI component for editing the hotspot/crop value of image fields in Studio, and can be seen inside the test studio.

Keys areas to consider with these changes would be any regression with interaction performance or browser support. Also being mindful of and potential performance bottlenecks that may have been missed.

### Testing

There is currently no unit test coverage on the logic of the canvas-based imagetool. New unit tests are in progress for the newly added utility functions and hooks, but due to the size of the PR it's probably best to review those separately while ensuring the new component works as expected here.

I've performed extensive testing in browser, if there are any edge cases that were not documented or covered in tests, please let me know!

### Notes for release

- The image hotspot/crop tool generally feels more responsive to input.
- You can now focus the hotspot or crop element and control with keyboard arrow keys (shift to increase the change, cmd/ctrl to resize instead of move).

### CODE REVIEW

**Review by skogsmaskin — APPROVED** [review]
This is great @jmswrnr!

So cool that it's focusable and keyboard usable now!

I have been testing the functionality, and it seems to work great. The performance seems to be on par with the old one, if not better.

One thing that is on my personal wishlist is that we could extract this functionality as  code that is reusable outside the ImageInput (exported from 'sanity', or even better a own package). 

We started forking this code in the Media Library project, where we also added flip and rotate to the tools last year. We have since removed image manipulation from the feature set we are offering in the ML at launch though.  We do plan to reintroduce it  at some point though, and then it should most definitely be a part of this common codebase.

We don't need to do it right now, but this is definitely too valuable not to be reusable.

**Comment by bjoerge** [comment]
This is fantastic @jmswrnr! I'm beyond happy to see this tool getting some love and finally becoming accessible ❤️ 

Code looks great, I covered the following during my testing:
- [x] Different device pixel ratios and viewport sizes
- [x] GIFs
- [x] Tiny images (upscales nicely)
- [x] Handles partial values (unsetting either `hotspot` and `crop` gracefully falls back to defaults)
- [x] Latest Firefox
- [x] Latest Chrome
- [x] Uploaded SVGs (upscales nicely, much better than today)
- [x] Safari (18.4)
- [x] Safari iOS (see scroll-issue below, otherwise working)

I was also pleasantly surprised to see that the keyboard shortcuts I instinctively tried (without looking at the code first) just turned out to work!

Also nice to see that some of the finer details from the original implementation are still preserved, in particular that resizing the crop rect so it becomes smaller than the hotspot it squeezes the hotspot ellipse to fill the crop rect, but if you then resize back without letting go of the mouse button, the hotspot ellipse grows back to its original size.

The only issue I could spot was that on Safari on iOS (I'm on 18.4), dragging the handles triggered the scroll event on the nearest scroll container, which seems to make the drag gesture loose the target element, see screen recording:
https://github.com/user-attachments/assets/22ce1f67-243b-4b68-b95a-f57465bee8dc

**Comment by jmswrnr** [comment]
Thanks @skogsmaskin !
> One thing that is on my personal wishlist is that we could extract this functionality as code that is reusable outside the ImageInput (exported from 'sanity', or even better, its own module).

That's a nice idea! I can certainly see the value in reusing some of the logic in other places, will explore following up with moving this into some package/export

Also thanks @bjoerge, appreciate the quick help with testing! I'll take a look at that iOS issue 👍

**Review by stipsan — DISMISSED** [review]
Stellar work 💖 

chefs-kiss-french-chef

File: packages/sanity/src/core/form/inputs/files/ImageToolInput/imagetool/ImageTool.tsx [hunk 1]
```diff
@@ -1,36 +1,74 @@
...
+
+    if (containerRef.current) {
+      setContainerSize({
+        width: containerRef.current.clientWidth,
+        height: containerRef.current.clientHeight,
+      })
+    }
+
+    return resizeObserver.observe(containerRef.current, updateSize)
+  }, [])
```
[the remark this answers is not in the document]
Comment (jmswrnr) [reply packages/sanity/src/core/form/inputs/files/ImageToolInput/imagetool/ImageTool.tsx]: Great spot! thanks for going into detail with this, applied changes in 94500a30bd1e2999547616a61e07542026421043 🙌

File: packages/sanity/src/core/form/inputs/files/ImageToolInput/imagetool/DragAwareCanvas.tsx [hunk 2]
```diff
@@ -1,182 +0,0 @@
...
-const StyledCanvas = styled.canvas`
-  display: block;
-  position: relative;
-  max-width: calc(100% - 0.5em); /* to prevent overlap with change bar */
-  max-height: calc(100% + 1em);
-  user-select: none;
-  // Enable only multi-finger panning and zooming within this element.
-  // This prevents single finger panning when manipulating drag handles,
-  // which can cause unwanted scrolling in the underlying document body.
-  touch-action: pinch-zoom;
```
[the remark this answers is not in the document]
Comment (jmswrnr) [reply packages/sanity/src/core/form/inputs/files/ImageToolInput/imagetool/DragAwareCanvas.tsx:160]: I did already have a `touch-action` on the element container, seems like the only way to truly prevent this scroll on iOS is with a directly attached event handler: 16bd13f4b92292de498e5918fd0a74968e71ee5b

Also applied some extras to remove input delay by opting out of that hold to highlight/magnify feature: 395907ca36d2d9897b307e0a92cc3c2721db0677

File: packages/sanity/src/core/form/inputs/files/ImageToolInput/imagetool/hooks/useKeyboardControls.ts [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+  }, [onChangeEnd, value])
+
+  // Clean up the debounce timeout on unmount
+  useEffect(() => {
+    return () => {
+      if (changeEndTimeoutRef.current !== null) {
+        window.clearTimeout(changeEndTimeoutRef.current)
+      }
+    }
+  }, [])
```
[the remark this answers is not in the document]
Comment (jmswrnr) [reply packages/sanity/src/core/form/inputs/files/ImageToolInput/imagetool/hooks/useKeyboardControls.ts]: `onChangeEnd` is used to commit saved changes, but appears that removing this `useEffect` has no negative impact that I can see (closing dialog with pending debounce) - removed in 226fc5e253b8a4a2b9f860ec2da6a9ce3a1e7a5f

**Comment by jmswrnr** [comment]
@stipsan - Appreciate the thorough code review and comments! everything should be addressed now. 🙌 

@bjoerge - The iOS interaction issue should also be resolved now. I'm happy to see you noticed those small interaction details that were preserved, those are great interactions I made sure to keep in the new implementation!

**Review by bjoerge — APPROVED** [review]
Confirming the iOS scroll issue is resolved 🎉

LGTM from my side!

File: packages/sanity/src/core/form/inputs/files/ImageToolInput/imagetool/DragAwareCanvas.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (stipsan) [reply packages/sanity/src/core/form/inputs/files/ImageToolInput/imagetool/DragAwareCanvas.tsx:160]: Gotcha, I'm not aware of a better pattern than touch-move no, iOS Safari is the new IE IMO 😅

**Review by stipsan — APPROVED** [review]

**Review by stipsan — APPROVED** [review]
Let's merge let's go! 🏎️💨 
giphy
